rush.md

```diff
@@ -0,0 +1,374 @@
+			new_batch = below(U)
+			remove from new_batch the contents of batch[0], batch[1], ..., batch[r-1]
+			sort new_batch using any canonical ordering
+			batch[r] = new_batch
+			let [U_1, U_2, ..., U_k] = new_batch
+			order = order ++ [U_1.data, U_2.data, ..., U_k.data]
+		else:
+			break
+	return order
+```
+ The `OrderData` algorithm is indeed quite simple: we choose a `Head` unit from each consecutive round (until it is no more possible) and generate a new batch of units, by taking all units below the head, and removing units from previous batches. The ordering is simply the data items from the units in subsequent batches, where each batch is ordered arbitrarily (yet using some deterministic ordering).  It is not hard to observe that `OrderData` has the desired monotonicity property as long as the black-box `Head(r, D)` is also "monotone", in the sense that:
+
+  **If `D` is a subset of `D'` and `Head(r, D)=Some(U)` then  `Head(r, D')=Some(U)`.**
+
+  We proceed to describing `Head(r, D)`
+
+#### 2.3.2 Description of Head.
+
+We will again describe `Head` as a procedure using another black-box: `Decide(U, D)` which takes a unit `U` and a dag `D` and outputs a value of type `Option<bool>`, so either a boolean "decision" `Some(b)` or `None`. We assume here that `Decide(U, D)` satisfies an analogous property to the one mentioned for `Head` above.
+
+```
+def Head(r, D):
+	let [U_1, U_2, ..., U_s] be the units in D of round r sorted according to some canonical order
+	// The order is allowed to depend on r and on Head(r-1, D)
+	for l = 1, 2, ..., s:
+		if Decide(U_l, D) == Some(b):
+			if b==true:
+				return U_l
+			else:
+				continue the loop
+		else:
+			return None
+```
+ To compute the head for a given round `r` we test all the units of this round in the dag `D` in some arbitrary canonical order (ordering by unit hashes is fine, but more sophisticated orderings here might give additional defense against delays caused by malicious nodes). The first unit that is decided `true` in this list is decided to be the head, however if `Decide()` outputs `None` at any time, `Head` must output `None` as well.
+
+#### 2.3.3 Description of Decide.
+Finally we describe the `Decide` procedure which is the last remaining piece to have a working consensus protocol. Again, `Decide` is based on a black-box `DecideVia(U, V, D)` which we describe subsequently:
+
+```
+def Decide(U, D):
+	if there exists any unit V in D such that DecideVia(U, V, D) == Some(b):
+		// In this case it is guaranteed that all bs will be the same
+		return Some(b)
+	else:
+		return None
+```
+
+To explain `DecideVia` we first need to define "virtual voting" (we refer to the [paper](https://arxiv.org/abs/1908.05156) for intuitions and clarifications) -- intuitively, the units that are "beyond U" (i.e. have higher round number) "vote" for whether we should decide `true` or `false` for `U`. It is good to think that this vote decides, whether "`U` is well known among nodes".
```

Comment:
Maybe you could emphasize that at this point the creator of a unit does not have any way to impact the vote and it is completely determined by the shape of the dag?

Reply:
Good point, added some comment on that.